src/SparseTypedFastBitSet.ts

```diff
@@ -0,0 +1,1380 @@
+  }
+
+  // Tries to add the value (Set the bit at index to true), return 1 if the
+  // value was added, return 0 if the value was already present
+  checkedAdd(index: number) {
+    const type = this.resize(index);
+    if (type === Type.ARRAY) {
+      const array = this.data;
+      if (this.arraySize === 0) {
+        array[this.arraySize++] = index;
+        return 1;
+      }
+      const offset = array.indexOf(index);
+      if (offset === -1 || offset >= this.arraySize) {
+        const biggest = array[0];
+        if (index > biggest) {
+          array[0] = index;
+          array[this.arraySize++] = biggest;
+        } else {
+          array[this.arraySize++] = index;
+        }
+        return 1;
+      }
+      return 0;
+    } else {
+      const words = this.data;
+      const word = words[index >>> 5];
+      const newword = word | (1 << index);
+      words[index >>> 5] = newword;
+      return ((newword ^ word) >>> index) as 0 | 1;
+    }
+  }
+
+  // Reduce the memory usage to a minimum
+  trim() {
+    const size = this.size();
+    if (this.arraySize === -1) {
+      const words = this.data;
+      let nl = words.length;
+      while (nl > 0 && words[nl - 1] === 0) {
+        nl--;
+      }
+      //sparse array will be smaller
+      if (size < nl) {
+        const newArray = new Uint32Array(size);
+        let pos = 0 | 0;
+        const c = words.length;
+        for (let k = 0; k < c; ++k) {
+          let w = words[k];
+          while (w != 0) {
+            const t = w & -w;
+            newArray[pos++] = (k << 5) + hammingWeight((t - 1) | 0);
+            w ^= t;
+          }
+        }
+
+        const first = newArray[0];
+        newArray[0] = newArray[size - 1];
+        newArray[size - 1] = first;
+
+        this.data = newArray;
+        this.arraySize = size;
+      } else {
+        this.data = words.slice(0, nl);
+      }
+    } else {
+      //dense bitset will be smaller
+      if (this.arraySize > 0 && this.data[0] >>> 5 < size) {
+        const count = (this.data[0] + 32) >>> 5;
+        const newWords = new Uint32Array(count);
+        for (let i = 0; i < this.arraySize; i++) {
+          const add = this.data[i];
+          newWords[add >>> 5] |= 1 << add;
+        }
+        this.data = newWords;
+        this.arraySize = -1;
+      } else {
+        this.data = this.data.slice(0, this.arraySize);
+      }
+    }
+  }
+
+  // Resize the bitset so that we can write a value at index
+  resize(index: number, capacity = 0): Type.ARRAY | Type.BITSET {
+    if (this.arraySize === -1) {
+      const words = this.data;
+      if (words.length << 5 > index) return Type.BITSET;
+      const count = (index + 32) >>> 5; // just what is needed
+      const newwords = new Uint32Array(count << 1);
+      newwords.set(words); // hopefully, this copy is fast
+      this.data = newwords;
+
+      return Type.BITSET;
+    } else {
+      const array = this.data;
+      if (this.arraySize + capacity < array.length) {
```

Comment:
So irrespective of 'number', if I call resize(number) and arraySize is smaller than array.length, then this function returns Type.ARRAY and does nothing else. Is that correct?

Reply:
Yip method returns the current datastructure used.

Save the calling method from doing the same check itself.

So if is basic array and capacity will return ARRAY if is a bitset and doesn't need to resize will return BITSET